# CAPSTONE PROJECT : MONEY-TOR [DATA SCIENCE]
- CDCC325D6Y0761 - Ignatius William Alvin
- CDCC325D6Y2656 - Jason Nathannael

## PERTANYAAN BISNIS
1. a
2. b
3. c
4. d

## IMPORT PACKAGES YANG DIPERLUKAN

In [111]:
# Import untuk Manipulasi Data
import pandas as pd
import numpy as np

# Import untuk Visualisasi Data
import matplotlib.pyplot as plt
import seaborn as sns

# Import untuk Tanggal
import datetime as dt

## DATA WRANGLING

### Gathering Data

In [112]:
# Membaca Dataset yang digunakan
transactions_df = pd.read_csv('Transactions.csv')
sellers_df = pd.read_csv('Sellers.csv')
products_df = pd.read_csv('Products.csv')
suppliers_df = pd.read_csv('Suppliers.csv')
logs_df = pd.read_csv('Logs.csv')

### Insight Gathering Data

- Pada tahap gathering data, dilakukan proses pengumpulan data dari berbagai sumber file dalam format CSV menggunakan library pandas dengan fungsi `pd.read_csv()`.

- Dataset yang digunakan terdiri dari beberapa tabel yang merepresentasikan entitas dalam sistem UMKM, seperti data transaksi (Transactions), data penjual (Sellers), data produk (Products), data penyuplai (Suppliers), serta data pendukung berupa riwayat aktivitas pembelian (Logs).

- Secara keseluruhan, dataset telah terstruktur dengan baik dan siap digunakan pada tahap analisis berikutnya, yaitu assessing data dan cleaning data.

### Assessing Data

In [113]:
# Menampilkan 5 baris teratas dari dataset Transactions menggunakan df.head
transactions_df.head()

,id_transaksi,id_seller,id_produk,tanggal,jenis_transaksi,kategori,qty,harga_satuan,total_harga,metode_pembayaran,event,diskon,status_transaksi
0,T00001,U006,P026,2026-05-28,pemasukan,penjualan_produk,1,85000,85000,QRIS,NaN,NaN,sukses
1,T00002,U006,P030,2026-05-15,pemasukan,penjualan_produk,1,55000,55000,Cash,NaN,NaN,sukses
2,T00003,U005,P025,2026-03-28,pemasukan,jasa_laundry,5,12000,60000,QRIS,NaN,NaN,sukses
3,T00004,U005,P021,2026-08-04,pemasukan,jasa_laundry,2,12000,24000,Transfer,NaN,NaN,sukses
4,T00005,U003,P012,2025-07-10,pengeluaran,restock_barang,3,18000,54000,Cash,NaN,NaN,sukses


In [114]:
# Menampilkan 5 baris teratas dari dataset Sellers menggunakan df.head
sellers_df.head()

,id_seller,nama_usaha,jenis_usaha,lokasi_usaha,tanggal_bergabung
0,U001,Warkop GenZ,Kafe,Bandung,2024-01-10
1,U002,Warung Madura,Retail,Bekasi,2024-01-15
2,U003,Makan Yuk,Kuliner,Yogyakarta,2025-02-01
3,U004,Sweet Bite Bakery,Bakery,Jakarta,2025-03-01
4,U005,LaundryKita,Laundry,Tangerang,2026-03-15


In [115]:
# Menampilkan 5 baris teratas dari dataset Products menggunakan df.head
products_df.head()

,id_produk,id_seller,id_supplier,nama_produk,kategori_produk,harga_jual,harga_modal,stok_awal,minimum_stok,status_produk,tanggal_input_produk
0,P001,U001,S001,Kopi Susu,Minuman,18000,9000,100,20,aktif,2024-01-15
1,P002,U001,S001,Americano,Minuman,15000,7000,100,20,aktif,2024-01-15
2,P003,U001,S001,Latte,Minuman,20000,10000,90,18,aktif,2024-01-15
3,P004,U001,S004,Croissant,Makanan,18000,9000,60,12,aktif,2024-01-16
4,P005,U001,S004,Cookies,Snack,12000,6000,70,15,aktif,2024-01-16


In [116]:
# Menampilkan 5 baris teratas dari dataset Suppliers menggunakan df.head
suppliers_df.head()

,id_supplier,nama_supplier,kategori_supplier,lokasi
0,S001,PT Kopi Nusantara,Biji Kopi & Minuman,Bandung
1,S002,CV Sembako Jaya,Sembako & FMCG,Bekasi
2,S003,UD Dapur Nusantara,Bahan Makanan,Yogyakarta
3,S004,UD Bakery Fresh,Bakery & Snack,Jakarta
4,S005,CV Kebutuhan Rumah,Household & Laundry,Tangerang


In [117]:
# Menampilkan 5 baris teratas dari dataset Logs menggunakan df.head
logs_df.head()

,id_log,id_produk,id_seller,tanggal_log,jenis_log,stok_sebelum,stok_sesudah,jumlah_perubahan,keterangan
0,L0304,P008,U002,2024-01-01,penyesuaian_stok,35,34,-1,Penyesuaian stok hasil audit
1,L0095,P005,U001,2024-01-02,penyesuaian_stok,53,63,10,Penyesuaian stok hasil audit
2,L0099,P017,U004,2024-01-06,stok_menipis,14,10,-4,Stok mendekati batas minimum
3,L0114,P008,U002,2024-01-09,penyesuaian_stok,20,12,-8,Penyesuaian stok hasil audit
4,L0324,P004,U001,2024-01-11,restock,34,60,26,Penambahan stok dari supplier


In [118]:
# Menampilkan informasi 5 dataset yag digunakan dengan df_info
transactions_df.info()
print("\n")

sellers_df.info()
print("\n")

products_df.info()
print("\n")

suppliers_df.info()
print("\n")

logs_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id_transaksi       2000 non-null   object 
 1   id_seller          2000 non-null   object 
 2   id_produk          2000 non-null   object 
 3   tanggal            2000 non-null   object 
 4   jenis_transaksi    2000 non-null   object 
 5   kategori           2000 non-null   object 
 6   qty                2000 non-null   int64  
 7   harga_satuan       2000 non-null   int64  
 8   total_harga        2000 non-null   int64  
 9   metode_pembayaran  2000 non-null   object 
 10  event              185 non-null    object 
 11  diskon             70 non-null     float64
 12  status_transaksi   2000 non-null   object 
dtypes: float64(1), int64(3), object(9)
memory usage: 203.3+ KB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 5 columns):
 #   Co

In [119]:
# Mengecek ukuran dataset menggunakan df.shape
print("Dataset Transactions:", transactions_df.shape)
print("Dataset Sellers:", sellers_df.shape)
print("Dataset Products:", products_df.shape)
print("Dataset Suppliers:", suppliers_df.shape)
print("Dataset Logs:", logs_df.shape)

Dataset Transactions: (2000, 13)
Dataset Sellers: (6, 5)
Dataset Products: (30, 11)
Dataset Suppliers: (8, 4)
Dataset Logs: (450, 9)


In [120]:
# Mengecek apakah ada duplikasi data pada dataset menggunakan df.duplicated().sum()
print("Jumlah duplikasi data pada dataset Transactions:", transactions_df.duplicated().sum())
print("Jumlah duplikasi data pada dataset Sellers:", sellers_df.duplicated().sum())
print("Jumlah duplikasi data pada dataset Products:", products_df.duplicated().sum())
print("Jumlah duplikasi data pada dataset Suppliers:", suppliers_df.duplicated().sum())
print("Jumlah duplikasi data pada dataset Logs:", logs_df.duplicated().sum())

Jumlah duplikasi data pada dataset Transactions: 0
Jumlah duplikasi data pada dataset Sellers: 0
Jumlah duplikasi data pada dataset Products: 0
Jumlah duplikasi data pada dataset Suppliers: 0
Jumlah duplikasi data pada dataset Logs: 0


In [121]:
# Mengecek apakah ada missing Value pada dataset Transactions
transactions_df.isnull().sum()

,0
id_transaksi,0
id_seller,0
id_produk,0
tanggal,0
jenis_transaksi,0
kategori,0
qty,0
harga_satuan,0
total_harga,0
metode_pembayaran,0


In [122]:
# Mengecek apakah ada missing Value pada dataset Sellers
sellers_df.isnull().sum()

,0
id_seller,0
nama_usaha,0
jenis_usaha,0
lokasi_usaha,0
tanggal_bergabung,0


In [123]:
# Mengecek apakah ada missing Value pada dataset Products
products_df.isnull().sum()

,0
id_produk,0
id_seller,0
id_supplier,0
nama_produk,0
kategori_produk,0
harga_jual,0
harga_modal,0
stok_awal,0
minimum_stok,0
status_produk,0


In [124]:
# Mengecek apakah ada missing Value pada dataset Suppliers
suppliers_df.isnull().sum()

,0
id_supplier,0
nama_supplier,0
kategori_supplier,0
lokasi,0


In [125]:
# Mengecek apakah ada missing Value pada dataset Logs
logs_df.isnull().sum()

,0
id_log,0
id_produk,0
id_seller,0
tanggal_log,0
jenis_log,0
stok_sebelum,0
stok_sesudah,0
jumlah_perubahan,0
keterangan,0


In [126]:
# Melihat deskripsi dataset Transactions menggunakan df.describe()
transactions_df.describe()

,qty,harga_satuan,total_harga,diskon
count,2000.00000,2000.000000,2000.000000,70.000000
mean,3.27400,25334.305500,72884.099500,11.928571
std,1.97126,23715.287348,92860.545846,5.535973
min,1.00000,2500.000000,0.000000,5.000000
25%,2.00000,12000.000000,24000.000000,6.250000
50%,3.00000,18000.000000,48000.000000,10.000000
75%,4.00000,22000.000000,85000.000000,15.000000
max,10.00000,95000.000000,750000.000000,20.000000


In [127]:
# Melihat deskripsi dataset Sellers menggunakan df.describe()
sellers_df.describe()

,id_seller,nama_usaha,jenis_usaha,lokasi_usaha,tanggal_bergabung
count,6,6,6,6,6
unique,6,6,6,5,6
top,U001,Warkop GenZ,Kafe,Bandung,2024-01-10
freq,1,1,1,2,1


In [128]:
# Melihat deskripsi dataset Products menggunakan df.describe()
products_df.describe()

,harga_jual,harga_modal,stok_awal,minimum_stok
count,30.000000,30.000000,30.000000,30.000000
mean,24933.333333,14600.000000,128.433333,19.966667
std,24667.699575,16298.614284,237.727993,22.421485
min,2500.000000,1500.000000,35.000000,7.000000
25%,12000.000000,6000.000000,50.000000,10.000000
50%,17000.000000,9000.000000,60.000000,13.500000
75%,20000.000000,12500.000000,90.000000,19.500000
max,95000.000000,65000.000000,999.000000,100.000000


In [129]:
# Melihat deskripsi dataset Suppliers menggunakan df.describe()
suppliers_df.describe()

,id_supplier,nama_supplier,kategori_supplier,lokasi
count,8,8,8,8
unique,8,8,8,7
top,S001,PT Kopi Nusantara,Biji Kopi & Minuman,Bandung
freq,1,1,1,2


In [130]:
# Melihat deskripsi dataset Logs menggunakan df.describe()
logs_df.describe()

,stok_sebelum,stok_sesudah,jumlah_perubahan
count,450.000000,450.000000,450.000000
mean,55.302222,65.682222,10.371111
std,121.785500,125.082775,18.116612
min,1.000000,0.000000,-15.000000
25%,15.000000,16.000000,-4.000000
50%,30.000000,41.500000,1.000000
75%,49.000000,68.750000,24.000000
max,982.000000,1017.000000,50.000000


### Insight Assessing Data

- Berdasarkan hasil assessing data menggunakan fungsi `info()`, `head()`, `duplicated()`, `isnull()`, `shape()`, dan `describe()`, seluruh dataset berhasil dimuat dengan baik dan memiliki struktur data yang cukup terorganisir untuk proses analisis lebih lanjut.

- Dataset Transactions merupakan dataset utama dengan jumlah data terbesar, yaitu sebanyak 2000 baris dan 13 kolom. Dataset ini memuat informasi transaksi UMKM seperti jenis transaksi, kategori transaksi, jumlah pembelian, metode pembayaran, diskon, dan status transaksi.

- Hasil pengecekan menunjukkan bahwa tidak ditemukan data duplikat pada seluruh dataset, sehingga kualitas data awal tergolong baik dan tidak memerlukan proses penghapusan duplicate data.

- Pada proses pengecekan missing value, ditemukan nilai kosong pada atribut event sebanyak 1815 data dan diskon sebanyak 1930 data pada dataset Transactions. Namun, kondisi tersebut masih tergolong wajar karena tidak seluruh transaksi menggunakan event promosi maupun diskon.

- Dataset Sellers, Products, Suppliers, dan Logs tidak memiliki missing value sehingga dapat dikategorikan memiliki kelengkapan data yang baik.

- Berdasarkan hasil pemeriksaan tipe data, beberapa atribut tanggal seperti tanggal, tanggal_bergabung, tanggal_input_produk, dan tanggal_log masih bertipe object sehingga perlu dikonversi menjadi tipe datetime pada tahap cleaning data agar lebih optimal dalam analisis berbasis waktu.

- Hasil statistik deskriptif menunjukkan bahwa rata-rata jumlah pembelian (qty) pada dataset transaksi berada pada kisaran 3 produk per transaksi dengan rata-rata total transaksi sebesar 72 ribu rupiah. Selain itu, nilai diskon rata-rata berada pada kisaran 11 - 12%.

- Pada dataset produk (Products), ditemukan variasi stok yang cukup besar dengan nilai maksimum stok awal mencapai 999 unit. Hal ini mengindikasikan adanya perbedaan skala penyimpanan stok antar produk.

- Dataset Logs menunjukkan adanya aktivitas perubahan stok berupa restock maupun pengurangan stok, yang terlihat dari nilai jumlah_perubahan yang dapat bernilai positif maupun negatif.

- Secara keseluruhan, dataset memiliki kualitas yang cukup baik dan siap dilanjutkan ke tahap cleaning data untuk menangani missing value serta penyesuaian tipe data sebelum dilakukan analisis dan visualisasi lebih lanjut.

### Cleaning Data

In [131]:
# =========================================================== #
# 1. Mengubah tipe data tanggal dari object menjadi datetime! #
# =========================================================== #

# Kolom yang akan dikonversi antara lain
#     transactions_df['tanggal']
#     sellers_df['tanggal_bergabung']
#     products_df['tanggal_input_produk']
#     logs_df['tanggal_log']

transactions_df['tanggal'] = pd.to_datetime(transactions_df['tanggal'])
sellers_df['tanggal_bergabung'] = pd.to_datetime(sellers_df['tanggal_bergabung'])
products_df['tanggal_input_produk'] = pd.to_datetime(products_df['tanggal_input_produk'])
logs_df['tanggal_log'] = pd.to_datetime(logs_df['tanggal_log'])

# Mengecek salah satu dataset yang sudah diubah kolom objectnya
transactions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   id_transaksi       2000 non-null   object        
 1   id_seller          2000 non-null   object        
 2   id_produk          2000 non-null   object        
 3   tanggal            2000 non-null   datetime64[ns]
 4   jenis_transaksi    2000 non-null   object        
 5   kategori           2000 non-null   object        
 6   qty                2000 non-null   int64         
 7   harga_satuan       2000 non-null   int64         
 8   total_harga        2000 non-null   int64         
 9   metode_pembayaran  2000 non-null   object        
 10  event              185 non-null    object        
 11  diskon             70 non-null     float64       
 12  status_transaksi   2000 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(3), object(8)
memory us

In [132]:
# ========================== #
# 2. Menangani Missing Value #
# ========================== #

# Missing Value yang ingin ditangani berasal dari dataset Transactions
#     Kolom Event  : diisii dengan tidak ada event
#     Kolom Diskon : diisi dengan 0

transactions_df['event'] = transactions_df['event'].fillna('Tidak Ada Event')
transactions_df['diskon'] = transactions_df['diskon'].fillna(0)

# Mengecek status kolom pada dataset transactions
transactions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   id_transaksi       2000 non-null   object        
 1   id_seller          2000 non-null   object        
 2   id_produk          2000 non-null   object        
 3   tanggal            2000 non-null   datetime64[ns]
 4   jenis_transaksi    2000 non-null   object        
 5   kategori           2000 non-null   object        
 6   qty                2000 non-null   int64         
 7   harga_satuan       2000 non-null   int64         
 8   total_harga        2000 non-null   int64         
 9   metode_pembayaran  2000 non-null   object        
 10  event              2000 non-null   object        
 11  diskon             2000 non-null   float64       
 12  status_transaksi   2000 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(3), object(8)
memory us

In [133]:
# ======================= #
# 3. Validasi Tidak Wajar #
# ======================= #

# Pada tahap assessing sebelumnya ditemukan: total_harga minimum = 0
# perlu memastikan apakah data tersebut: valid atau merupakan anomali/error.

# 3a. Mengecek transaksi dengan total harga = 0
transactions_df[transactions_df['total_harga'] == 0]


,id_transaksi,id_seller,id_produk,tanggal,jenis_transaksi,kategori,qty,harga_satuan,total_harga,metode_pembayaran,event,diskon,status_transaksi
31,T00032,U001,P005,2026-07-05,pemasukan,penjualan_produk,5,13799,0,Transfer,Tidak Ada Event,0.0,batal
79,T00080,U003,P014,2026-12-17,pemasukan,penjualan_produk,4,12000,0,Cash,Tidak Ada Event,0.0,batal
117,T00118,U004,P017,2025-12-10,pengeluaran,promo_diskon,4,8000,0,Transfer,Tidak Ada Event,0.0,batal
124,T00125,U002,P010,2025-10-12,pemasukan,penjualan_produk,8,2500,0,Cash,Tidak Ada Event,0.0,batal
132,T00133,U004,P017,2026-12-28,pemasukan,penjualan_produk,1,8000,0,Cash,Tidak Ada Event,0.0,batal
...,...,...,...,...,...,...,...,...,...,...,...,...,...
1839,T01840,U001,P002,2026-07-04,pemasukan,penjualan_produk,5,17250,0,QRIS,Tidak Ada Event,0.0,batal
1903,T01904,U003,P011,2026-04-13,pengeluaran,restock_barang,2,20000,0,Cash,Tidak Ada Event,0.0,batal
1931,T01932,U004,P019,2025-03-30,pemasukan,penjualan_produk,3,25299,0,QRIS,Ramadan,15.0,batal
1969,T01970,U005,P021,2026-06-08,pemasukan,jasa_laundry,5,12000,0,Cash,Tidak Ada Event,0.0,batal


In [134]:
# ======================= #
# 3. Validasi Tidak Wajar #
# ======================= #

# Pada tahap 3a sebelumnya terlihat bahwa:
# total_harga = 0 bukan error acak, karena seluruh transaksi tersebut memiliki: status_transaksi = batal

# 3b. Mengecek apakah semua total_harga == 0 memang memiliki status_transaksi == 'batal'
transactions_df[transactions_df['total_harga'] == 0]['status_transaksi'].value_counts()

,count
status_transaksi,
batal,71


In [135]:
# ========================================= #
# 4. Mengecekk konsistensi Data Kategorikal #
# ========================================= #

# memastikan tidak ada penulisan kategori yang berbeda seperti:
#     Cash
#     cash
#     CASH

print("Metode Pembayaran:")
print(transactions_df['metode_pembayaran'].unique())

print("\nStatus Transaksi:")
print(transactions_df['status_transaksi'].unique())

print("\nJenis Transaksi:")
print(transactions_df['jenis_transaksi'].unique())

print("\nKategori Transaksi:")
print(transactions_df['kategori'].unique())

print("\nStatus Produk:")
print(products_df['status_produk'].unique())

Metode Pembayaran:
['QRIS' 'Cash' 'Transfer']

Status Transaksi:
['sukses' 'batal']

Jenis Transaksi:
['pemasukan' 'pengeluaran']

Kategori Transaksi:
['penjualan_produk' 'jasa_laundry' 'restock_barang' 'promo_diskon'
 'pembelian_bahan']

Status Produk:
['aktif']


In [136]:
# ================================== #
# 5a.Validasi Perhitungan Total Harga #
# ================================== #

# memastikan rumus transaksi benar.
#    total_harga = qty × harga_satuan

invalid_total = transactions_df[
    (transactions_df['qty'] * transactions_df['harga_satuan'] != transactions_df['total_harga']) &
    (transactions_df['status_transaksi'] != 'batal')
]

invalid_total

,id_transaksi,id_seller,id_produk,tanggal,jenis_transaksi,kategori,qty,harga_satuan,total_harga,metode_pembayaran,event,diskon,status_transaksi
26,T00027,U001,P003,2024-03-14,pemasukan,penjualan_produk,2,20000,32000,QRIS,Ramadan,20.0,sukses
44,T00045,U003,P011,2026-12-12,pengeluaran,promo_diskon,2,23000,43700,Transfer,12.12,5.0,sukses
49,T00050,U002,P008,2024-03-28,pemasukan,penjualan_produk,4,19200,72960,Cash,Ramadan,5.0,sukses
82,T00083,U003,P013,2026-02-24,pemasukan,penjualan_produk,4,7200,27360,QRIS,Ramadan,5.0,sukses
89,T00090,U001,P003,2024-04-03,pemasukan,penjualan_produk,3,20000,51000,Cash,Ramadan,15.0,sukses
...,...,...,...,...,...,...,...,...,...,...,...,...,...
1843,T01844,U004,P018,2025-03-07,pengeluaran,restock_barang,5,17250,77625,Cash,Ramadan,10.0,sukses
1884,T01885,U004,P018,2025-03-17,pemasukan,penjualan_produk,3,15000,40500,Cash,Ramadan,10.0,sukses
1890,T01891,U001,P005,2026-02-20,pemasukan,penjualan_produk,2,13799,26218,QRIS,Ramadan,5.0,sukses
1898,T01899,U001,P001,2026-03-08,pemasukan,penjualan_produk,2,20700,39330,Transfer,Ramadan,5.0,sukses


In [137]:
# ================================================= #
# 5b.Validasi Perhitungan Total Harga dengan diskon #
# ================================================= #

# Pada tahap 5a sebelumnya terlihat bahwa:
# qty x harga_satuan != total_harga yang terjadi pada 68 transaksi sukses.
# Artinya transaksi sudah memperhitungkan diskon.

# Validasi total_harga dengan memperhitungkan diskon
validasi_diskon = transactions_df[
    (
        (
            transactions_df['qty'] *
            transactions_df['harga_satuan'] *
            (1 - transactions_df['diskon'] / 100)
        ).round(0) != transactions_df['total_harga']
    )
    &
    (transactions_df['status_transaksi'] != 'batal')
]

validasi_diskon

,id_transaksi,id_seller,id_produk,tanggal,jenis_transaksi,kategori,qty,harga_satuan,total_harga,metode_pembayaran,event,diskon,status_transaksi
887,T00888,U003,P011,2026-03-07,pemasukan,penjualan_produk,3,26999,64797,Transfer,Ramadan,20.0,sukses
1136,T01137,U003,P014,2026-02-22,pemasukan,penjualan_produk,5,16199,72895,Transfer,Ramadan,10.0,sukses
1239,T01240,U004,P016,2025-03-28,pemasukan,penjualan_produk,4,13799,44156,Transfer,Ramadan,20.0,sukses
1437,T01438,U003,P012,2025-03-01,pemasukan,penjualan_produk,5,24299,109345,Transfer,Ramadan,10.0,sukses


### Insight Cleaning Data

- Pada tahap cleaning data, dilakukan proses pembersihan dan penyesuaian dataset agar data lebih siap digunakan pada tahap analisis dan visualisasi.

- Proses cleaning pertama dilakukan dengan mengubah tipe data pada beberapa atribut tanggal dari tipe object menjadi datetime, seperti pada kolom tanggal, tanggal_bergabung, tanggal_input_produk, dan tanggal_log. Perubahan ini bertujuan agar data dapat digunakan secara optimal dalam analisis berbasis waktu

- Selanjutnya dilakukan penanganan missing value pada dataset Transactions, khususnya pada kolom event dan diskon. Nilai kosong pada kolom event diisi dengan label "Tidak Ada Event" sedangkan kolom diskon diisi dengan nilai 0, karena tidak seluruh transaksi menggunakan promosi maupun potongan harga.

- Setelah proses penanganan missing value dilakukan, seluruh atribut pada dataset transaksi berhasil memiliki jumlah data lengkap tanpa nilai kosong.

- Tahap cleaning berikutnya dilakukan dengan memvalidasi transaksi yang memiliki total_harga = 0. Hasil pemeriksaan menunjukkan bahwa seluruh transaksi tersebut memiliki status_transaksi = batal, sehingga data dinilai valid dan tetap dipertahankan karena masih memiliki nilai analisis terhadap tingkat pembatalan transaksi.

- Selain itu, dilakukan pengecekan konsistensi data kategorikal pada atribut seperti metode_pembayaran, status_transaksi, jenis_transaksi, kategori, dan status_produk. Hasilnya menunjukkan bahwa seluruh kategori telah konsisten tanpa adanya perbedaan penulisan data.

- Proses validasi perhitungan transaksi juga dilakukan untuk memastikan kesesuaian antara qty, harga_satuan, diskon, dan total_harga. Sebagian besar data telah sesuai dengan perhitungan diskon yang diterapkan, sedangkan beberapa selisih kecil ditemukan akibat proses pembulatan desimal pada sistem transaksi sehingga data tetap dianggap valid.

- Secara keseluruhan, hasil cleaning data menunjukkan bahwa dataset telah memiliki kualitas data yang baik, konsisten, serta siap digunakan pada tahap exploratory data analysis (EDA) dan visualisasi data selanjutnya.

## Exploratory Data Analysis (EDA)

### Hal yang diexplore adalah:
Pada tahap analisis ini akan berpusat pada hubungan timbal-balik antara Pemasaran dan Inventori:

Dari pemasaran (Pelanggan & Penjualan): Mencari tahu produk apa yang paling laku, kapan waktu ramai pembeli, dan taktik promosi/pembayaran apa yang paling disukai pelanggan untuk mendongkrak pendapatan.

Dari inventori (Stok & Operasional): Memastikan kelancaran pasokan agar tidak kehilangan potensi penjualan akibat kehabisan stok serta memantau kesehatan pengeluaran bisnis melalui efisiensi penyuplai.

In [138]:
# Memastikan hanya menganalisis transaksi sukses
sukses_transaksi_df = transactions_df[transactions_df['status_transaksi'] == 'sukses'].copy()

print(f"Total data transaksi berhasil: {len(sukses_transaksi_df)} baris")

Total data transaksi berhasil: 1929 baris


In [139]:
# Total Pemasukan & Pengeluaran
revenue_vs_expense = sukses_transaksi_df.groupby('jenis_transaksi')['total_harga'].sum().reset_index()
print("Total Pemasukan & Pengeluaran (Sukses)")
print(revenue_vs_expense)
print("\n")

# Rasio Transaksi Sukses & Batal
status_ratio = transactions_df['status_transaksi'].value_counts(normalize=True) * 100
print("Rasio Status Transaksi (%)")
print(status_ratio.round(2))
print("\n")

# Transaksi Berdasarkan Kategori
category_analysis = sukses_transaksi_df.groupby('kategori').agg(
    jumlah_transaksi=('id_transaksi', 'count'),
    total_nilai=('total_harga', 'sum')
).sort_values(by='total_nilai', ascending=False)
print("Analisis Berdasarkan Kategori Transaksi")
print(category_analysis)

Total Pemasukan & Pengeluaran (Sukses)
  jenis_transaksi  total_harga
0       pemasukan    123496558
1     pengeluaran     22271641


Rasio Status Transaksi (%)
status_transaksi
sukses    96.45
batal      3.55
Name: proportion, dtype: float64


Analisis Berdasarkan Kategori Transaksi
                  jumlah_transaksi  total_nilai
kategori                                       
penjualan_produk              1358    113382358
jasa_laundry                   291     10114200
pembelian_bahan                103      9208242
restock_barang                  91      6828962
promo_diskon                    86      6234437


In [140]:
#  Menampilkan Top 5 Produk Terlaris berdasarkan Kuantitas (Qty)
top_products_qty = sukses_transaksi_df.groupby('id_produk')['qty'].sum().reset_index()
top_products_qty = top_products_qty.merge(products_df[['id_produk', 'nama_produk', 'kategori_produk']], on='id_produk')
top_products_qty = top_products_qty.sort_values(by='qty', ascending=False).head(5)

print("Top 5 Produk Terlaris (Kuantitas)")
print(top_products_qty[['nama_produk', 'kategori_produk', 'qty']].to_string(index=False))
print("\n")

# Metode Pembayaran Terpopuler
payment_trends = sukses_transaksi_df.groupby('metode_pembayaran').agg(
    jumlah_pengguna=('id_transaksi', 'count'),
    total_nominal=('total_harga', 'sum')
).sort_values(by='jumlah_pengguna', ascending=False)

print("Preferensi Metode Pembayaran")
print(payment_trends)
print("\n")

# Pengaruh Diskon terhadap Pembelian Kuantitas (Qty)
promo_impact = sukses_transaksi_df.groupby('event').agg(
    rata_rata_qty=('qty', 'mean'),
    rata_rata_diskon=('diskon', 'mean'),
    total_omset=('total_harga', 'sum')
).sort_values(by='total_omset', ascending=False)

print("Dampak Event Promosi & Diskon")
print(promo_impact)

Top 5 Produk Terlaris (Kuantitas)
     nama_produk kategori_produk  qty
       Beras 5kg         Sembako  381
Minyak Goreng 1L         Sembako  366
  Gula Pasir 1kg         Sembako  357
     Kopi Sachet         Minuman  339
      Mie Instan         Makanan  311


Preferensi Metode Pembayaran
                   jumlah_pengguna  total_nominal
metode_pembayaran                                
QRIS                           719       43651756
Cash                           656       58262349
Transfer                       554       43854094


Dampak Event Promosi & Diskon
                 rata_rata_qty  rata_rata_diskon  total_omset
event                                                        
Tidak Ada Event       3.221841          0.000000    133292587
Ramadan               3.912281          6.578947      7288631
Natal                 3.000000          0.000000      2908895
Tahun Baru            4.333333          0.000000      1877187
12.12                 3.285714          5.000000     

In [141]:
# Memastikan kolom tanggal bertipe datetime
sukses_transaksi_df['tanggal'] = pd.to_datetime(sukses_transaksi_df['tanggal'])

# Analisis Hari Ter ramai dalam Seminggu
sukses_transaksi_df['hari_pembelian'] = sukses_transaksi_df['tanggal'].dt.day_name()
order_hari = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

traffic_hari = sukses_transaksi_df.groupby('hari_pembelian').agg(
    jumlah_transaksi=('id_transaksi', 'count'),
    total_omset=('total_harga', 'sum')
).reindex(order_hari)

print("Analisis Trafik & Penjualan Berdasarkan Hari")
print(traffic_hari)

Analisis Trafik & Penjualan Berdasarkan Hari
                jumlah_transaksi  total_omset
hari_pembelian                               
Monday                       266     19083190
Tuesday                      287     22100740
Wednesday                    281     21232700
Thursday                     277     19228240
Friday                       290     20654656
Saturday                     278     22326628
Sunday                       250     21142045


In [142]:
#  Hitung total produk terjual dari transaksi sukses
total_terjual = sukses_transaksi_df.groupby('id_produk')['qty'].sum().reset_index()
total_terjual.columns = ['id_produk', 'total_terjual']

# Hitung Sisa Stok Aktual (Stok Awal - Total Terjual)
stok_aktual_df = products_df.merge(total_terjual, on='id_produk', how='left')
stok_aktual_df['total_terjual'] = stok_aktual_df['total_terjual'].fillna(0)
stok_aktual_df['stok_sekarang'] = stok_aktual_df['stok_awal'] - stok_aktual_df['total_terjual']

# Deteksi Produk yang Mendekati/Di Bawah Batas Minimum Stok
urgent_restock_aktual = stok_aktual_df[stok_aktual_df['stok_sekarang'] <= stok_aktual_df['minimum_stok']]

print(f"Deteksi Kritis: Ada {len(urgent_restock_aktual)} Produk Butuh Restock Segera")
if len(urgent_restock_aktual) > 0:
    print(urgent_restock_aktual[['nama_produk', 'stok_awal', 'total_terjual', 'stok_sekarang', 'minimum_stok']].to_string(index=False))
else:
    print("Semua stok produk aman di atas batas minimum.")

Deteksi Kritis: Ada 28 Produk Butuh Restock Segera
     nama_produk  stok_awal  total_terjual  stok_sekarang  minimum_stok
       Kopi Susu        100            228           -128            20
       Americano        100            212           -112            20
           Latte         90            195           -105            18
       Croissant         60            146            -86            12
         Cookies         70            191           -121            15
       Beras 5kg         40            381           -341            10
Minyak Goreng 1L         60            366           -306            15
  Gula Pasir 1kg         50            357           -307            12
      Mie Instan        120            311           -191            30
     Kopi Sachet        100            339           -239            25
     Nasi Goreng         70            174           -104            15
     Ayam Geprek         60            260           -200            12
    Es Teh Ma

In [143]:
# Menghubungkan top 5 produk terlaris dengan sisa stok real-time saat ini
cross_analysis = top_products_qty[['id_produk', 'nama_produk', 'qty']].merge(
    stok_aktual_df[['id_produk', 'stok_sekarang', 'minimum_stok']],
    on='id_produk'
)

print("Analisis Lintas Divisi: Kesehatan Stok untuk Produk Terlaris")
print(cross_analysis.to_string(index=False))
print("\n")

# Performa Penjualan Berdasarkan Nama Usaha Seller
seller_performance = sukses_transaksi_df.merge(sellers_df, on='id_seller')
seller_summary = seller_performance.groupby('nama_usaha').agg(
    total_penjualan=('total_harga', 'sum'),
    jumlah_transaksi=('id_transaksi', 'count')
).sort_values(by='total_penjualan', ascending=False)

print("Performa Penjualan Berdasarkan Nama Usaha")
print(seller_summary)

Analisis Lintas Divisi: Kesehatan Stok untuk Produk Terlaris
id_produk      nama_produk  qty  stok_sekarang  minimum_stok
     P006        Beras 5kg  381           -341            10
     P007 Minyak Goreng 1L  366           -306            15
     P008   Gula Pasir 1kg  357           -307            12
     P010      Kopi Sachet  339           -239            25
     P009       Mie Instan  311           -191            30


Performa Penjualan Berdasarkan Nama Usaha
                   total_penjualan  jumlah_transaksi
nama_usaha                                          
Warung Madura             43231010               296
Thrift Hub                43165000               318
Warkop GenZ               17025764               325
Makan Yuk                 15808827               330
Sweet Bite Bakery         15166398               329
LaundryKita               11371200               331


## Insight Exploratory Data Analysis (EDA)
- 1. Temuan Operasional Gudang

- Jika dilihat dari angka, bisnis pada UMKM secara finansial sangat sehat karena Pemasukan (Rp123,4 Juta) jauh lebih besar daripada Pengeluaran (Rp22,2 Juta). Tingkat pembatalan transaksi juga sangat rendah, yaitu hanya 3,55%. Namun, ada masalah di bagian gudang.

- Fenomena Overselling: Ada 28 produk andalan yang memiliki sisa stok minus ratusan. Contoh paling parah adalah Beras 5kg (minus 341), Minyak Goreng (minus 306), dan Kopi Sachet (minus 239).

- Dampak: Pelanggan membeli barang yang "tidak ada" di Etalase, atau penjual berhasil menjualnya tetapi mencatatnya secara inden. Jika ini terus dibiarkan tanpa sinkronisasi, penjual akan kehilangan pembeli karna tidak adanya ketersediaan barang.

- 2. Insight Pemasaran & Perilaku Pelanggan
- A. Top 5 Produk Terlaris didominasi total oleh kategori Sembako (Beras, Minyak, Gula) dan Makanan/Minuman cepat saji (Kopi, Mie Instan). Hal ini sejalan dengan performa toko, di mana Warung Madura (Sembako) dan Thrift Hub (Fashion) menjadi penyumbang masing-masing Rp43 Juta+. Kombinasi dari kedua toko ini menguasai hampir 70% dari total omset bisnis.

- B. Pola Transaksi Harian
Transaksi harian cenderung stabil di angka 250–290 transaksi per hari sepanjang minggu. Tapi, hari Sabtu dan Selasa menjadi hari paling menghasilkan dengan omset menembus Rp22 Juta+. Tingginya omset di hari Selasa menunjukkan adanya perilaku belanja tengah pekan, kemungkinan besar oleh pelanggan tetap yang menyetok ulang kebutuhan mereka.

- C. QRIS & Cash
- QRIS adalah Metode ini paling banyak dipilih dengan 719 pengguna, pelanggan sangat menyukai kepraktisan digital.
Meskipun penggunanya lebih sedikit (656 pengguna), total uang tunai yang masuk mencapai Rp58,2 Juta (jauh lebih besar dari QRIS yang hanya Rp43,6 Juta). Artinya, transaksi bernominal besar atau belanja grosir masih dominan menggunakan uang tunai.

- D. Efektivitas Promosi
- Secara total omset, hari biasa yang Tidak Ada Event menyumbang Rp133 Juta karena durasinya yang panjang sepanjang tahun. Tapi, jika melihat Rata-rata Qty per Transaksi, event khusus sangat mendongkrak keranjang belanja pelanggan. Pada event 11.11, rata-rata pelanggan membeli 5,3 produk per transaksi, dan pada Ramadan sebesar 3,9 produk di atas hari biasa yang hanya 3,2 produk.

3. Pemasaran & Inventori
- Sinkronisasi Produk Terlaris: Beras 5kg adalah produk terlaris (terjual 381 qty) tetapi stoknya defisit parah hingga minus 341.

- Event: Karena event seperti Ramadan dan 11.11 terbukti menaikkan jumlah pembelian barang per transaksi hingga di atas 20%, inventori tidak boleh memakai standar stok normal. Batas minimum stock harus dinaikkan 1.5x lipat khusus untuk produk Sembako dan Kopi sebulan sebelum event dimulai sebagai stok pengaman.

## Visualization & Explanatory Analysis

### Pertanyaan bisnis 1

In [144]:
# kode pertanyaan bisnis 1

### Insight pertanyaan bisnis 1
- a
- b

### Pertanyaan bisnis 2

In [145]:
# kode pertanyaan bisnis 2

### Insight pertanyaan bisnis 2
- a
- b

## Analisis Lanjutan

In [146]:
# kode untuk membuat file untuk keperluan dashboard streamlit

## KESIMPULAN PROSES ANALISA

1. a
2. b
3. c